File: Chapter 14/Python/section_14_6_instrumentation_logging/ports/payment_port.py
"""
Outbound port definition for third-party billing interactions.
Decouples core processing execution from concrete network clients.

Book listing: com.ecommerce.order.ports.PaymentPort — Listing 14.1
"""

from abc import ABC, abstractmethod


class PaymentPort(ABC):
    """Abstract base class defining the payment processing boundary contract."""

    @abstractmethod
    def process(self, amount: float) -> bool:
        """Process a payment for the given amount."""
        ...

File: Chapter 14/JS/section_14_6_instrumentation_logging/ports/paymentPort.js
/**
 * Outbound port definition for third-party billing interactions.
 * Decouples core processing execution from concrete network clients.
 *
 * Book listing: com.ecommerce.order.ports.PaymentPort — Listing 14.1
 */

/**
 * PaymentPort defines the boundary contract.
 * In JavaScript, this is expressed as a structural contract (duck typing):
 * any object with a `process(amount)` method satisfies the port.
 */
export class PaymentPort {
  /**
   * Process a payment for the given amount.
   * @param {number} amount - The payment amount.
   * @returns {boolean} True if payment succeeded.
   */
  process(amount) {
    throw new Error("PaymentPort.process() must be implemented by a concrete adapter.");
  }
}